application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/UpgradeExtensionHandler.java

```diff
@@ -0,0 +1,210 @@
+
+/**
+ * Upgrades an extension from a namespace to the last compatible version and sends a notification. The notification will
+ * be displayed only for users that subscribed to it and disabled the System filter since it is send by superadmin user.
+ * 
+ * @version $Id$
+ * @since 1.17
+ */
+@Component(roles = UpgradeExtensionHandler.class)
+@Singleton
+public class UpgradeExtensionHandler
+{
+    /**
+     * The id of application-licensing-licensor-api module.
+     */
+    protected static final String LICENSOR_API_ID = "com.xwiki.licensing:application-licensing-licensor-api";
+
+    @Inject
+    protected DocumentAccessBridge documentAccessBridge;
+
+    @Inject
+    protected JobExecutor jobExecutor;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private ExtensionRepositoryManager extensionRepositoryManager;
+
+    /** The default factory for creating event objects. */
+    @Inject
+    private ObservationManager observationManager;
+
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> currentDocumentReferenceResolver;
+
+    @Inject
+    private ContextualLocalizationManager localization;
+
+    /**
+     * Try upgrading an extension inside a namespace to the last compatible version.
+     * 
+     * @param installedExtension the already installed extension
+     * @param namespace the namespace in which the extension is installed
+     */
+    public void tryUpgradeExtensionToLastVersion(InstalledExtension installedExtension, String namespace)
+    {
+        ExtensionId installedExtensionId = installedExtension.getId();
+        // Use the reversed list of versions since the last compatible version is targeted.
+        List<Version> versions = getInstallableVersions(installedExtensionId);
+
+        for (Version version : versions) {
+            ExtensionId toInstallExtensionId = new ExtensionId(installedExtensionId.getId(), version);
+            Boolean upgradeDone = true;
+            try {
+                installExtension(toInstallExtensionId, namespace);
+
+                String doneUpgradeMessage = this.localization.getTranslationPlain("licensor.notifications.event.done",
+                    installedExtension.getName(), installedExtensionId.getVersion().getValue(),
+                    toInstallExtensionId.getVersion().getValue());
+
+                this.observationManager.notify(new ExtensionAutoUpgradedEvent(), LICENSOR_API_ID, doneUpgradeMessage);
+            } catch (JobException | InterruptedException e) {
+                String failedUpgradeMessage = this.localization.getTranslationPlain(
+                    "licensor.notifications.event.failed", installedExtension.getName(),
+                    installedExtensionId.getVersion().getValue(), toInstallExtensionId.getVersion().getValue());
+
+                this.observationManager.notify(new ExtensionAutoUpgradedEvent(), LICENSOR_API_ID, failedUpgradeMessage);
```

Comment:
It doesn't feel right to send the same event for success and failure. If I listen to ``ExtensionAutoUpgradedEvent`` then I expect this event to be triggered after the extension was automatically upgraded successfully. For instance, I don't expect ``DocumentDeletedEvent`` to be sent if the document failed to be deleted. At most (but still strange) you should be able to know from the event object whether the operation succeeded or not (asserting the message string is not an option). But it may be cleaner to have separate events (I don't know what's the best practice for notifications regarding this).